source/manual/alerts/email-alert-api-unprocessed-work.html.md

```diff
@@ -0,0 +1,57 @@
+---
+owner_slack: "#govuk-2ndline"
+title: 'Email Alert API: Unprocessed work'
+section: Icinga alerts
+subsection: Email alerts
+layout: manual_layout
+parent: "/manual.html"
+last_reviewed_on: 2020-10-05
+review_in: 6 months
+---
+
+> Sometimes we lose work due to [a flaw with the Sidekiq queueing system](https://github.com/mperham/sidekiq/wiki/Problems-and-Troubleshooting#my-sidekiq-process-is-crashing-what-do-i-do). At 30 minute intervals a job, [RecoverLostJobsWorker], will try to requeue work that has not been processed [within an hour](https://github.com/alphagov/email-alert-api/blob/2f3931ac1ca25fe8c79b2405af98d1de55e1d47b/app/workers/recover_lost_jobs_worker/unprocessed_check.rb#L13).
+
+This alert indicates that Email Alert API has work that has not been processed in the generous amount of time we expect it to have been - see [here](https://github.com/alphagov/email-alert-api/tree/master/app/workers/metrics_collection_worker) for the timings for each alert.
+
+* **[`unprocessed content changes`](https://github.com/alphagov/email-alert-api/blob/master/app/workers/process_content_change_worker.rb)**. This means we're not generating emails for subscribers with "immediate" frequency subscriptions in response to [a change in some content] on GOV.UK.
+
+* **[`unprocessed messages`](https://github.com/alphagov/email-alert-api/blob/master/app/workers/process_message_worker.rb)**. This means we're not generating emails for subscribers with "immediate" frequency subscriptions in response to [a custom message].
+
+* **`incomplete digest runs`**. This could be due to a failure in any of three workers:
+
+   * [\[Daily/Weekly\]DigestInitiatorWorker](https://github.com/alphagov/email-alert-api/blob/a656389b1abdd46226ca37c1682c318f1c2eafee/app/workers/daily_digest_initiator_worker.rb) generates a DigestRunSubscriber work item for each subscriber.
+   * [DigestEmailGenerationWorker](https://github.com/alphagov/email-alert-api/blob/a656389b1abdd46226ca37c1682c318f1c2eafee/app/workers/digest_email_generation_worker.rb) does the work of generating the digest email for a specific subscriber.
+   * [DigestRunCompletionMarkerWorker](https://github.com/alphagov/email-alert-api/blob/a656389b1abdd46226ca37c1682c318f1c2eafee/app/workers/digest_run_completion_marker_worker.rb) periodically scans all the work items to see if the run is complete.
+
+Things to check:
+
+* Check [Sentry] for errors.
+
+* Check the [Sidekiq dashboard] for worker failures.
+
+* Check [Kibana] for errors - use ```@fields.worker: <worker class>``` for the query.
+
+* Manually retry any pending jobs from a console: `Sidekiq::RetrySet.new.retry_all`.
```

Comment:
This feels inconsistent with the list being an action rather than something to check. It'd feel more appropriate to suggest they check the [Sidekiq monitoring](https://docs.publishing.service.gov.uk/manual/sidekiq.html#sidekiq-web-aka-sidekiq-monitoring) to see if there are large queues or the item is in the retry set.

It also feels a potentially rather big action doing the whole retry set.

Reply:
This is something I found useful when attempting to fix an issue with Content Data API, where the backoff had grown large, and I wanted more immediate feedback. On reflection, I think it was pretty niche, and the main scenarios are:

- Broken code that we fix pretty quickly
- A transient error that goes away quickly

In both scenarios, the retry schedule should mean the alert goes away by itself, and we don't need to manually flush the retry set.

I've removed this suggestion (the one above covers the monitoring part).